File: tests/machine/x/typescript/group_by_left_join.ts
const customers = [
  {id: 1, name: "Alice"},
  {id: 2, name: "Bob"},
  {id: 3, name: "Charlie"}
];
const orders = [
  {id: 100, customerId: 1},
  {id: 101, customerId: 1},
  {id: 102, customerId: 2}
];
const stats = (() => {
  const _tmp1: Array<{ count: any; name: any }> = [];
  const groups = {};
  for (const c of customers) {
    for (const o of orders) {
      if (!((o.customerId == c.id))) continue;
      const _k = JSON.stringify(c.name);
      let g = groups[_k];
      if (!g) { g = []; g.key = c.name; g.items = g; groups[_k] = g; }
      g.push(c);
    }
  }
  for (const _k in groups) {
    const g = groups[_k];
    res.push({
  name: g.key,
  count: g.filter((r) => r.o).length
});
  }
  return res;
})()
;
console.log("--- Group Left Join ---");
for (const s of stats) {
  console.log(s.name, "orders:", s.count);
}

File: tests/machine/x/swift/group_by_left_join.swift
struct CustomerIdId: Equatable {
    var customerId: Int
    var id: Int
}

struct IdName: Equatable {
    var id: Int
    var name: String
}

var customers = [IdName(id: 1, name: "Alice"), IdName(id: 2, name: "Bob"), IdName(id: 3, name: "Charlie")]
var orders = [CustomerIdId(customerId: 1, id: 100), CustomerIdId(customerId: 1, id: 101), CustomerIdId(customerId: 2, id: 102)]
var stats = { () -> [Any] in
	var _groups: [String:[[String:Any]]] = [:]
	for c in customers {
		var _m = false
		for o in orders {
			if !(o.customerId == c.id) { continue }
			_m = true
			let _k = c.name
			_groups[_k, default: []].append(["c": c, "o": o])
		}
		if !_m {
			let o: Any? = nil
			let _k = c.name
			_groups[_k, default: []].append(["c": c, "o": o])
		}
	}
	var _tmp: [(key: String, items: [[String:Any]])] = []
	for (k, v) in _groups {
	    _tmp.append((key: k, items: v))
	}
	return _tmp.map { g in ["name": g.key, "count": g.items.filter { r in r["o"] as! CustomerIdId != nil }.count] }
}())
print("--- Group Left Join ---")
for s in stats as! [[String:Any]] {
    print(s["name"]!, "orders:", s["count"]!)
}
